worldedit-core/src/main/java/com/fastasyncworldedit/core/queue/implementation/ApplyTask.java

```diff
@@ -0,0 +1,241 @@
+            ConcurrentMap<Thread, ThreadState<F>> stateCache,
+            boolean full,
+            boolean[] faweExceptionReasonsUsed
+    ) {
+
+    }
+
+    private static final class ThreadState<F extends Filter> {
+
+        private final SingleThreadQueueExtent queue;
+        private final F filter;
+        private ChunkFilterBlock block;
+
+        private ThreadState(SingleThreadQueueExtent queue, F filter) {
+            this.queue = queue;
+            this.filter = filter;
+        }
+
+    }
+
+    ApplyTask(
+            final Region region,
+            final F filter,
+            final ParallelQueueExtent parallelQueueExtent,
+            final boolean full, final boolean[] faweExceptionReasonsUsed
+    ) {
+        this.commonState = new CommonState<>(
+                filter,
+                region.clone(), // clone only once, assuming the filter doesn't modify that clone
+                parallelQueueExtent,
+                new ConcurrentHashMap<>(),
+                full,
+                faweExceptionReasonsUsed
+        );
+        this.before = null;
+        final BlockVector3 minimumPoint = region.getMinimumPoint();
+        this.minChunkX = minimumPoint.x() >> 4;
+        this.minChunkZ = minimumPoint.z() >> 4;
+        final BlockVector3 maximumPoint = region.getMaximumPoint();
+        this.maxChunkX = maximumPoint.x() >> 4;
+        this.maxChunkZ = maximumPoint.z() >> 4;
+        this.shift = INITIAL_REGION_SHIFT;
+
+    }
+
+    private ApplyTask(
+            final CommonState<F> commonState,
+            final ApplyTask<F> before,
+            final int minChunkX,
+            final int maxChunkX,
+            final int minChunkZ,
+            final int maxChunkZ,
+            final int higherShift
+    ) {
+        this.commonState = commonState;
+        this.minChunkX = minChunkX;
+        this.maxChunkX = maxChunkX;
+        this.minChunkZ = minChunkZ;
+        this.maxChunkZ = maxChunkZ;
+        this.before = before;
+        this.shift = Math.max(0, higherShift - SHIFT_REDUCTION);
+    }
+
+    @Override
+    protected void compute() {
+        if (this.minChunkX != this.maxChunkX || this.minChunkZ != this.maxChunkZ) {
+            ApplyTask<F> subtask = null;
+            int minRegionX = this.minChunkX >> this.shift;
+            int minRegionZ = this.minChunkZ >> this.shift;
+            int maxRegionX = this.maxChunkX >> this.shift;
+            int maxRegionZ = this.maxChunkZ >> this.shift;
+            // This task covers multiple regions. Create one subtask per region
+            for (int regionX = minRegionX; regionX <= maxRegionX; regionX++) {
+                for (int regionZ = minRegionZ; regionZ <= maxRegionZ; regionZ++) {
+                    if (ForkJoinTask.getSurplusQueuedTaskCount() > Settings.settings().QUEUE.PARALLEL_THREADS) {
+                        // assume we can do a bigger batch of work here - the other threads are busy for a while
```

Comment:
I would replace can with should. Whilst it's a trade-off between cross-thread mask access, I would still call higher parallelism the "wanted" outcome